src/gc.c

```diff
@@ -1402,344 +1321,901 @@ STATIC_INLINE uintptr_t gc_read_stack(void *_addr, uintptr_t offset,
-                          uintptr_t lb, uintptr_t ub, int d)
+JL_NORETURN NOINLINE void gc_assert_datatype_fail(jl_ptls_t ptls, jl_datatype_t *vt,
+                                                  gc_mark_sp_t sp)
 {
-    while (s != NULL) {
-        jl_value_t ***rts = (jl_value_t***)(((void**)s) + 2);
-        size_t nroots = gc_read_stack(&s->nroots, offset, lb, ub);
-        size_t nr = nroots >> 1;
-        if (nroots & 1) {
-            for (size_t i = 0; i < nr; i++) {
-                void **slot = (void**)gc_read_stack(&rts[i], offset, lb, ub);
-                void *obj = (void*)gc_read_stack(slot, offset, lb, ub);
-                if (obj != NULL) {
-                    gc_push_root(ptls, obj, d);
-                }
-            }
+    jl_printf(JL_STDOUT, "GC error (probable corruption) :\n");
+    gc_debug_print_status();
+    jl_(vt);
+    gc_debug_critical_error();
+    gc_mark_loop_unwind(ptls, sp, 0);
+    abort();
+}
+
+// This stores the label address in the mark loop function.
+// We can't directly store that to a global array so we need some hack to get that.
+// See the call to `gc_mark_loop` in init with a `NULL` `ptls`.
+void *gc_mark_label_addrs[gc_mark_nlabels];
+
+// Double the mark stack (both pc and data) with the lock held.
+static void NOINLINE gc_mark_stack_resize(jl_gc_mark_cache_t *gc_cache, gc_mark_sp_t *sp)
+{
+    char *old_data = gc_cache->data_stack;
+    void **pc_stack = sp->pc_start;
+    size_t stack_size = (char*)sp->pc_end - (char*)pc_stack;
+    JL_LOCK_NOGC(&gc_cache->stack_lock);
+    gc_cache->data_stack = (char*)realloc(old_data, stack_size * 2 * sizeof(gc_mark_sp_t));
+    sp->data += gc_cache->data_stack - old_data;
+
+    sp->pc_start = gc_cache->pc_stack = (void**)realloc(pc_stack, stack_size * 2 * sizeof(void*));
+    gc_cache->pc_stack_end = sp->pc_end = sp->pc_start + stack_size * 2;
+    sp->pc += sp->pc_start - pc_stack;
+    JL_UNLOCK_NOGC(&gc_cache->stack_lock);
+}
+
+// Push a work to the stack. The type of the work is marked with `pc` and the data needed
+// is in `data` and is of size `data_size`.
+// The `sp` keeps track of the current stack pointer and will be updated on return.
+// If there isn't enough space on the stack anymore, the stack will be resized with the stack
+// lock held. The caller should invalidate any local cache of the stack addresses that's not
+// in `gc_cache` or `sp`
+// The caller is also responsible for increasing `pc`.
+STATIC_INLINE void gc_mark_stack_push(jl_gc_mark_cache_t *gc_cache, gc_mark_sp_t *sp,
+                                      void *pc, void *data, size_t data_size, int inc)
+{
+    assert(data_size <= sizeof(gc_mark_data_t));
+    if (__unlikely(sp->pc == sp->pc_end))
+        gc_mark_stack_resize(gc_cache, sp);
+    *sp->pc = pc;
+    memcpy(sp->data, data, data_size);
+    if (inc) {
+        sp->data += data_size;
+        sp->pc++;
+    }
+}
+
+// Check if the reference is non-NULL and atomically set the mark bit.
```

Comment:
[Atomic operation](https://github.com/JuliaLang/julia/blob/d7c00faca48484adfb46db2b5b7fdfdb114a710f/src/gc.c#L542) kind of atomic?